### PR TITLE
Simplify the version bump process

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Simplify the version bump process.

- Use substitution hook in docs to update CDN URL.
  - In `dev` or `prerelease` versions, it will point to `dev` url
  - In `stable` versions, it will point it `vX.Y.Z` url
- Introduce a version bump script `tools/bump_version.py`.

~- Use `bump2version` to bump Python versions (continuing from #2460)~
~- Use `npm version` to bump JS versions~
~- Update `run_docker` script to use latest deployed pre-built docker image when `--pre-built` option is specified.~

### Checklists

- [x] Add new / update outdated documentation

### CODE REVIEW

**Comment by ryanking13** [comment]
> * Use `bump2version` to bump Python versions (continuing from [Add bump2version #2460](https://github.com/pyodide/pyodide/pull/2460))
> * Use `npm version` to bump JS versions

We probably need to write our own script to deal with these two at once.

I struggled with `bump2version` for hours and my conclusion is that it does not perfectly fit our use case because Python and Javascript have slightly different versioning schemes in prereleases. For example, We use `0.20.0a1` for Python and `0.20.0-alpha.1` for JavaScript. Also `bump2version` doesn't seem to handle version upgrades correctly for dev and prerelease versions.

**Review by rth — COMMENTED** [review]
Thanks for spending time on this!

> We probably need to write our own script to deal with these two at once.
> 
> I struggled with bump2version for hours and my conclusion is that it does not perfectly fit our use case because Python and Javascript have slightly different versioning schemes in prereleases

Well the current situation where there is one tool to bump the js version and another tool to bump versions everywhere else also doesn't sound too bad. In any case one has to review the diff before committing and worse case scenario the JS versions needs adjusting a bit.

File: .bumpversion.cfg [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[bumpversion]
+current_version = 0.21.0.dev0
+files = src/py/pyodide/__init__.py src/py/setup.cfg pyodide-build/setup.cfg docs/conf.py
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply .bumpversion.cfg:3]: bump2version rewrites the config file, but the pre-commit doesn't like its formatting. So I moved it to a separate config file and make pre-commit to ignore it.

Anyway, I am thinking of writing a short script instead of using `bump2version`, because `bump2version` doesn't seem to work intuitively to me when dealing with prerelease of dev versions.

File: .bumpversion.cfg [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+parse = ^
+	(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)
+	((?P<pre>a|b|rc)(?P<preversion>\d+))?
+	(\.(?P<dev>dev)(?P<devversion>\d+))?
+serialize = 
+	{major}.{minor}.{patch}{pre}{preversion}.{dev}{devversion}
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply .bumpversion.cfg:9]: Yes.

File: run_docker [hunk 3]
```diff
@@ -62,7 +61,15 @@ do
           echo "WARNING: will use the env var PYODIDE_DOCKER_IMAGE=${PYODIDE_DOCKER_IMAGE},
           the flag --pre-built has no effect"
         fi
-        DEFAULT_PYODIDE_DOCKER_IMAGE="pyodide/pyodide:${PYODIDE_PREBUILT_IMAGE_TAG}"
+        LATEST_PREBUILT_DOCKER_IMAGE_TAG=$(wget -q \
+          https://registry.hub.docker.com/v1/repositories/pyodide/pyodide/tags -O - |
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply run_docker:65]: Alright, I'll revert this change :)

[the remark this answers is not in the document]
Comment (rth) [reply run_docker:65]: Apart from this everything else LGTM.

**Comment by ryanking13** [comment]
I dropped `bump2version` and instead wrote a Python script which handles both Python and JS version bump.
It's not very short but I think it will reduce our mistake.

<details>
<summary>Output (0.22.0)</summary>

```sh
$ ./tools/bump_version.py --new-version 0.22.0 --dry-run
[*] Diff of '/workspaces/pyodide/src/py/pyodide/__init__.py':

@@ -53 +53 @@
-__version__ = "0.21.0.dev0"
+__version__ = "0.22.0"

[*] Diff of '/workspaces/pyodide/src/py/setup.cfg':

@@ -3 +3 @@
-version = 0.21.0.dev0
+version = 0.22.0

[*] Diff of '/workspaces/pyodide/pyodide-build/setup.cfg':

@@ -3 +3 @@
-version = 0.21.0.dev0
+version = 0.22.0

[*] Diff of '/workspaces/pyodide/docs/conf.py':

@@ -18 +18 @@
-pyodide_version = "0.21.0.dev0"
+pyodide_version = "0.22.0"

[*] Diff of '/workspaces/pyodide/run_docker':

@@ -5 +5 @@
-PYODIDE_PREBUILT_IMAGE_TAG="0.20.0"
+PYODIDE_PREBUILT_IMAGE_TAG="0.22.0"

[*] Diff of '/workspaces/pyodide/docs/project/about.md':

@@ -48 +48 @@
-  version      = {0.20.0},
+  version      = {0.22.0},

[*] Diff of '/workspaces/pyodide/src/js/package.json':

@@ -3 +3 @@
-  "version": "0.21.0-dev.0",
+  "version": "0.22.0",

[*] Diff of '/workspaces/pyodide/src/js/package-lock.json':

@@ -3 +3 @@
-  "version": "0.21.0-dev.0",
+  "version": "0.22.0",
@@ -9 +9 @@
-      "version": "0.21.0-dev.0",
+      "version": "0.22.0",
```

</details>

<details>
<summary>Output (0.22.0a1)</summary>

```sh
$ ./tools/bump_version.py --new-version 0.22.0a1 --dry-run
[*] Diff of '/workspaces/pyodide/src/py/pyodide/__init__.py':

@@ -53 +53 @@
-__version__ = "0.21.0.dev0"
+__version__ = "0.22.0a1"

[*] Diff of '/workspaces/pyodide/src/py/setup.cfg':

@@ -3 +3 @@
-version = 0.21.0.dev0
+version = 0.22.0a1

[*] Diff of '/workspaces/pyodide/pyodide-build/setup.cfg':

@@ -3 +3 @@
-version = 0.21.0.dev0
+version = 0.22.0a1

[*] Diff of '/workspaces/pyodide/docs/conf.py':

@@ -18 +18 @@
-pyodide_version = "0.21.0.dev0"
+pyodide_version = "0.22.0a1"

[*] /workspaces/pyodide/run_docker: Skipped (not targeting a core version)
[*] /workspaces/pyodide/docs/project/about.md: Skipped (not targeting a core version)
[*] Diff of '/workspaces/pyodide/src/js/package.json':

@@ -3 +3 @@
-  "version": "0.21.0-dev.0",
+  "version": "0.22.0-alpha.1",

[*] Diff of '/workspaces/pyodide/src/js/package-lock.json':

@@ -3 +3 @@
-  "version": "0.21.0-dev.0",
+  "version": "0.22.0-alpha.1",
@@ -9 +9 @@
-      "version": "0.21.0-dev.0",
+      "version": "0.22.0-alpha.1",
```

</details>

File: docs/conf.py [hunk 4]
```diff
@@ -14,7 +14,13 @@
 # -- Project information -----------------------------------------------------
 
 project = "Pyodide"
-copyright = "2019-2021, Pyodide contributors and Mozilla"
+copyright = "2019-2022, Pyodide contributors and Mozilla"
+pyodide_version = "0.21.0.dev0"
+
+if any([(dev_ver in pyodide_version) for dev_ver in ("dev", "a", "b", "rc")]):
+    CDN_URL = "https://cdn.jsdelivr.net/pyodide/dev/full/"
```
Comment (hoodmane) [inline docs/conf.py:21]: Is this right? For instance for an alpha or beta release we would want to take the "else" block. I think you maybe just want `if "dev" in pyodide_version:`

File: setup.cfg [hunk 5]
```diff
@@ -1,42 +1,19 @@
...
-    .artifacts
+	build
+	cpython
+	emsdk
+	.artifacts
 addopts =
-    --doctest-modules
-    --ignore="packages/matplotlib/src"
+	--doctest-modules
+	--ignore="packages/matplotlib/src"
```
Comment (hoodmane) [inline setup.cfg:9]: Maybe better to revert this change? In most files we use spaces not tabs.

File: tools/bump_version.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        else:
+            raise ValueError(
+                f"'{file}' contains invalid version: expected '{current_version}' but found '{version}'"
+            )
+
+    show_diff(content, new_content, file)
+
+    if dry_run:
+        return lambda: None
+    return functools.partial(file.write_text, new_content)
```
Comment (hoodmane) [inline tools/bump_version.py:164]:
```suggestion
    return new_content
```

File: tools/bump_version.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+    if match is None:
+        raise ValueError(f"Unabled to detect version string: {target.file}")
+
+    return match.groupdict()["version"]
+
+
+def update_version(
+    target: Target, current_version: str, new_version: str, dry_run: bool = False
+) -> Callable:
```
Comment (hoodmane) [inline tools/bump_version.py:129]:
```suggestion
def generate_updated_content(
    target: Target, current_version: str, new_version: str
) -> Callable:
```

File: tools/bump_version.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    if re.match(PYTHON_VERSION_REGEX, new_version) is None:
+        raise ValueError(f"Invalid new version: {new_version}")
+
+    # We want to update files in all-or-nothing strategy,
+    # so we keep the queue of update functions
+    update_queue = []
+    for target in PYTHON_TARGETS:
+        current_version = parse_current_version(target)
+        update_queue.append(
+            update_version(target, current_version, new_version, dry_run=args.dry_run)
```
Comment (hoodmane) [inline tools/bump_version.py:204]:
```suggestion
            [target.file, generate_updated_content(target, current_version, new_version)]
```

File: tools/bump_version.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    for target in JS_TARGETS:
+        current_version = parse_current_version(target)
+        update_queue.append(
+            update_version(
+                target,
+                current_version,
+                python_version_to_js_version(new_version),
+                dry_run=args.dry_run,
+            )
+        )
```
Comment (hoodmane) [inline tools/bump_version.py:215]:
```suggestion
        update_queue.append(
            [target.file, generate_updated_content(
                target,
                current_version,
		js_version)
            ]
        )
```

File: tools/bump_version.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+    # We want to update files in all-or-nothing strategy,
+    # so we keep the queue of update functions
+    update_queue = []
+    for target in PYTHON_TARGETS:
+        current_version = parse_current_version(target)
+        update_queue.append(
+            update_version(target, current_version, new_version, dry_run=args.dry_run)
+        )
+    for target in JS_TARGETS:
```
Comment (hoodmane) [inline tools/bump_version.py:206]:
```suggestion
    js_version = python_version_to_js_version(new_version)
    for target in JS_TARGETS:
```

File: tools/bump_version.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        current_version = parse_current_version(target)
+        update_queue.append(
+            update_version(
+                target,
+                current_version,
+                python_version_to_js_version(new_version),
+                dry_run=args.dry_run,
+            )
+        )
+
```
Comment (hoodmane) [inline tools/bump_version.py:216]:
```suggestion
    if args.dry_run:
        return
```

File: tools/bump_version.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    for update_func in update_queue:
+        update_func()
```
Comment (hoodmane) [inline tools/bump_version.py:218]:
```suggestion
    for [file, content] in update_queue:
        file.write_text(content)
```

**Review by hoodmane — APPROVED** [review]
Thanks @ryanking13 this looks really nice. Ideally we should probably add some tests for `bump_version.py`. I also gave some mild refactoring suggestions, I think it will make testing easier.

**Comment by hoodmane** [comment]
Well the changes I suggested aren't quite right, but something vaguely similar. I guess you need a new way to indicate that a file need not be updated.

File: docs/conf.py (hunk 4 above)
Comment (ryanking13) [reply docs/conf.py:21]: You are right. I thought that we don't deploy alpha versions to CDN but we do.

File: setup.cfg (hunk 5 above)
Comment (ryanking13) [reply setup.cfg:9]: Yeah, this file was updated by `bump2version`.

**Comment by ryanking13** [comment]
Thanks for the reviews!
